src/core/ext/filters/client_channel/lb_policy.h

```diff
@@ -325,8 +326,20 @@ class LoadBalancingPolicy : public InternallyRefCounted<LoadBalancingPolicy> {
   /// Data passed to the UpdateLocked() method when new addresses and
   /// config are available.
   struct UpdateArgs {
-    ServerAddressList addresses;
+    /// A list of addresses, or an error indicating a failure to obtain the
+    /// list of addresses.
+    absl::StatusOr<ServerAddressList> addresses;
+    /// The LB policy config.
     RefCountedPtr<Config> config;
+    /// A human-readable note providing context about the name resolution that
+    /// provided this update.  LB policies may wish to include this message
+    /// in RPC failure status messages.  For example, if the update has an
+    /// empty list of addresses, this message might say "no DNS entries
+    /// found for <name>".
+    std::string resolution_note;
```

Comment:
I'm a little worried that this `resolution_note` field will make it confusing as to which error should be propagated (between the statuses of addresses and config, and this field).

It looks like there are no resolvers actually setting this field, and also `ClientChannel::CreateOrUpdateLbPolicyLocked` is not forwarding this field over to LB policy `UpdateArgs`

Meanwhile, LB policies like pick_first are referencing this field in their error message for "empty address list".

For such error messages, wouldn't all relevent debug info just be in the `address` field's absl status?

Reply:
The `resolution_note` is intended to be used specifically in the case where the resolver returns an empty list of addresses instead of an error.

The idea here is that there are cases where a resolver may intentionally want to return an empty address list without indicating an error, because it's going to use an LB policy that gets the backend addresses via some out-of-band control plane.  However, not all LB policies will work without being given addresses; in particular, leaf policies like pick_first or round_robin require addresses.  So if someone accidentally configures a resolver to return an empty list of address with an LB policy that requires addresses, the LB policy needs some additional information in order to provide a useful status message for RPC failures.

We don't currently have any resolvers that need to provide this information.  The only one we have that will return an empty list of addresses is the xds resolver, but it always configures the LB policy tree to get the backend addresses via EDS, so its result will never wind up being sent to a policy that requires a non-empty address list.

However, there are cases that I think we should ultimately support that we're just not handling correctly today.  For example, I think in the c-ares resolver, when an A/AAAA query fails, we should be checking the error code from that failure: if we get NXDOMAIN, that just means that there were no matching A/AAAA entries, so the resolver should return an empty address list instead of returning an error.  (In contrast, if the query fails with something like SRVFAIL, then we should return an error.)  This would allow, for example, using grpclb without any fallback addresses, by populating SRV records but no A records.  And in the future, when we make the resolver and LB policy APIs public, people will be able to write their own policies that do this sort of thing.

Internally, see go/grpc-client-channel-principles-revamp for more details.

You're right that we should be propagating this in `ClientChannel::CreateOrUpdateLbPolicyLocked()` -- thanks for catching that!  I've fixed that, and I've added a TODO about a similar propagation that needs to happen in the xds_cluster_resolver LB policy.

Reply:
If we expect `resolution_note` to be used if and only if `addresses` is an empty address list, what do you think about changing the type of the `addresses` field to something like `absl::variant<absl::StatusOr<ServerAddressList>, std::string>` ?

The second entry could be set by resolvers when they want to return an empty list with no error, but fill in a message. That way, it's easier to be sure that error messages aren't accidentally getting dropped.

> However, there are cases that I think we should ultimately support that we're just not handling correctly today. For example, I think in the c-ares resolver, when an A/AAAA query fails, we should be checking the error code from that failure: if we get NXDOMAIN, that just means that there were no matching A/AAAA entries, so the resolver should return an empty address list instead of returning an error. (In contrast, if the query fails with something like SRVFAIL, then we should return an error.) This would allow, for example, using grpclb without any fallback addresses, by populating SRV records but no A records. And in the future, when we make the resolver and LB policy APIs public, people will be able to write their own policies that do this sort of thing.

Interesting. For this specific case, I personally think that NXDOMAIN should still actually be considered an error, because depending on the specific DNS serving infrastructure in use, NXDOMAIN can sometimes be a transient error (e.g. b/151958298). I can see the usefulness of the ability do things like that in general though.

Reply:
I think using `absl::variant<>` in this case would be far more cumbersome.  In the majority of cases, we just want to pass along the address list as-is, without caring if it's empty or not.  But using `absl::variant<>` would force every caller to check the value, which IMHO is fairly ugly.

Ultimately, I think `resolution_note` is optional, and it's not the end of the world if it's not populated.  It's useful for improving error messages, but it's not actually required for things to work right, so I'd rather not make the core functionality here more cumbersome just to avoid a case where it may be desirable but isn't present.

With regard to NXDOMAIN, I do understand that there are probably existing cases in the wild where that error code will be caused by transient failures, but I would argue that any such case (including the bug you referenced) is a misbehavior on the part of the DNS serving infrastructure.  I'd probably be willing to have a channel arg to control whether or not the c-ares resolver interprets NXDOMAIN as an empty address list or as a transient failure, and we can argue about what the right default for that channel arg would be.  But in principle, I assert that NXDOMAIN should not be considered a transient failure.

But in any case, we can debate that particular issue more later. :)